File: local/src/lib/shoppinglist.js
import { AWORMap } from "$lib/crdt/awormap.js";
import { Product } from "$lib/product.js";

export class ShoppingList {
    #id = null;
    #products = null;
    #name = null;
    #remove = null;

    /**
     *
     * @param {String} id - Shopping list identifier
     */
    constructor(id, name, remove = false) {
        this.#id = id;
        this.#products = new AWORMap();
        this.#name = name;
        this.#remove = remove;
    }

    setName(name) {
        this.#name = name;
    }

    getName() {
        return this.#name;
    }

    setRemove(remove) {
        this.#remove = remove;
    }

    getRemove() {
        return this.#remove;
    }

    /**
     * Get shopping list identifier
     *
     * @returns shopping list identifier
     */
    getID() {
        const res = this.#id;
        return res;
    }

    /**
     * Get shopping list products
     *
     * @returns Products AWORMap
     */
    getProducts() {
        const res = this.#products;
        return res;
    }

    /**
     * Get specific product
     *
     * @param {String} product_name
     *
     * @returns Products list
     */
    getProduct(product_name) {
        return this.#products.get(product_name);
    }

    /**
     * Add new product to shopping list
     *
     * @param {Product} product
     */
    addProduct(product) {
        const exists = this.#products.get(product.getName());
        if (!exists) this.#products.add(product.getName(), product);
        this.incProduct(product.getName());
    }

    /**
     * Remove product from shopping list
     *
     * @param {Product} product
     */
    removeProduct(product_name) {
        this.#products.rm(product_name);
    }

    /**
     * Increment product amount
     *
     * @param {String} product_name
     * @param {Number} tosum
     */
    incProduct(product_name, tosum = 1) {
        const product = this.#products.get(product_name);
        if (product) product.inc(tosum);
    }

    /**
     * Decrement product amount
     *
     * @param {String} product_name
     * @param {Number} tosum
     */
    decProduct(product_name, tosum = 1) {
        const product = this.#products.get(product_name);
        if (product && product.value() > 0) product.dec(tosum);
        if (product && product.value() <= 0 && this.#remove)
            this.removeProduct(product_name);
    }

    /**
     * Merge shopping lists
     *
     * @param {ShoppingList} shoppinglist
     */
    join(shoppinglist) {
        // join counters
        for (const [key, value] of this.#products.getMap()) {
            const new_product = shoppinglist.getProduct(key[0]);
            if (new_product) {
                value.join(new_product);
            }
        }

        // join products
        this.#products.join(shoppinglist.getProducts());
    }

    /**
     * Get product from shopping list
     *
     * @param {String} product_name
     * @returns Product
     */
    get(product_name) {
        return this.#products.get(product_name);
    }

    /**
     * Create a Shopping List from a JSON object
     *
     * @param {JSON} json
     */

    fromJSON(json) {
        this.#id = json.id;
        this.#name = json.name;
        this.#remove = json.remove;
        this.#products = new AWORMap();
        // TODO: Iterate over products here
        this.#products.fromJSON(json.id, json.products);
    }

    toJSON() {
        const res = {
            id: this.#id,
            name: this.#name,
            remove: this.#remove,
            products: this.#products.toJSON(),
        };
        return res;
    }

    toFrontendJSON() {
        const res = {
            id: this.#id,
            name: this.#name,
            products: this.#products.toFrontendJSON(),
        };
        return res;
    }

    setUUID(uuid) {
        for (const [key, value] of this.#products.getMap()) {
            value.setUUID(uuid);
        }
    }
}

File: cloud/src/main/java/server/model/ShoppingList.java
package server.model;

import org.json.JSONObject;

import server.model.crdts.AWORMap;
import server.model.utils.Pair;

import java.util.Iterator;
import java.util.Map;

public class ShoppingList {
    private String id;
    private String name;
    private AWORMap products;

    /**
     * Constructor
     *
     * @param id Shopping list identifier
     */
    public ShoppingList(String id, String name) {
        this.id = id;
        this.name = name;
        this.products = new AWORMap();
    }

    /**
     * Get shopping list identifier
     *
     * @return Shopping list identifier
     */
    public String getId() {
        return id;
    }

    /**
     * Get shopping list display name
     *
     * @return Shopping list name
     */
    public String getName() {
        return name;
    }

    /**
     * Get shopping list products
     *
     * @return Products list
     */
    public AWORMap getProducts() {
        return products;
    }

    /**
     * Add a new product to the shopping list
     *
     * @param product Product to be added
     */
    public void addProduct(Product product) {
        products.add(product.getName(), product);
    }

    /**
     * Remove a product from the shopping list
     *
     * @param productName Name of the product to be removed
     */
    public void removeProduct(String productName) {
        products.rm(productName);
    }

    /**
     * Increment product amount
     *
     * @param {String} product_name
     * @param {Number} tosum
     */
    public void incProduct(String productName, int toSum) {
        Product product = this.products.get(productName);
        if (product != null && product.value() >= 0) {
            product.inc(toSum);
        }
    }

    /**
     * Decrement product amount
     *
     * @param {String} product_name
     * @param {Number} tosum
     */
    public void decProduct(String productName, int toSum) {
        Product product = this.products.get(productName);
        if (product != null && product.value() > 0) {
            product.dec(toSum);
        }
    }

    /**
     * Merge two shopping lists
     *
     * @param shoppingList Shopping list to be merged
     */
    public void join(ShoppingList shoppingList) {
        // Join products
        products.join(shoppingList.products);

        // Join counters
        for (Map.Entry<Pair<String, Integer>, Product> entry : products.entries()) {
            Product newProduct = shoppingList.get(entry.getKey().getKey()); // gets the element_id
            if (newProduct != null) {
                entry.getValue().join(newProduct);
            }
        }
    }

    /**
     * Get a product from the shopping list
     *
     * @param productName Name of the product to retrieve
     * @return Product
     */
    public Product get(String productName) {
        return products.get(productName);
    }

    /**
     * Create a Shopping List from a JSON object
     *
     * @param jsonString JSON string received
     */
    public void fromJSON(String jsonString) {
        JSONObject json = new JSONObject(jsonString);
        this.id = json.getString("id");
        this.name = json.getString("name");
        this.products = new AWORMap();
        this.products.fromJSON(json.getString("id"), json.getJSONObject("products"));
    }

    /*
    public String toJSON() {
        JSONObject res = new JSONObject();
        res.put("id", this.id);
        res.put("name", this.name);
        res.put("products", new JSONObject(this.products.toJSON()));
        return res.toString();
    }
    */
}
